File: MPChartLib/src/main/java/com/github/mikephil/charting/renderer/BubbleChartRenderer.java

package com.github.mikephil.charting.renderer;

import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint.Style;
import android.graphics.drawable.Drawable;

import com.github.mikephil.charting.animation.ChartAnimator;
import com.github.mikephil.charting.data.BubbleData;
import com.github.mikephil.charting.data.BubbleEntry;
import com.github.mikephil.charting.highlight.Highlight;
import com.github.mikephil.charting.interfaces.dataprovider.BubbleDataProvider;
import com.github.mikephil.charting.interfaces.datasets.IBubbleDataSet;
import com.github.mikephil.charting.utils.MPPointF;
import com.github.mikephil.charting.utils.Transformer;
import com.github.mikephil.charting.utils.Utils;
import com.github.mikephil.charting.utils.ViewPortHandler;

import java.util.List;

/**
 * Bubble chart implementation: Copyright 2015 Pierre-Marc Airoldi Licensed
 * under Apache License 2.0 Ported by Daniel Cohen Gindi
 */
public class BubbleChartRenderer extends BarLineScatterCandleBubbleRenderer {

    protected BubbleDataProvider mChart;

    public BubbleChartRenderer(BubbleDataProvider chart, ChartAnimator animator,
                               ViewPortHandler viewPortHandler) {
        super(animator, viewPortHandler);
        mChart = chart;

        renderPaint.setStyle(Style.FILL);

        highlightPaint.setStyle(Style.STROKE);
        highlightPaint.setStrokeWidth(Utils.convertDpToPixel(1.5f));
    }

    @Override
    public void initBuffers() {

    }

    @Override
    public void drawData(Canvas c) {

        BubbleData bubbleData = mChart.getBubbleData();

        for (IBubbleDataSet set : bubbleData.getDataSets()) {

            if (set.isVisible())
                drawDataSet(c, set);
        }
    }

    private float[] sizeBuffer = new float[4];
    private float[] pointBuffer = new float[2];

    protected float getShapeSize(float entrySize, float maxSize, float reference, boolean normalizeSize) {
        final float factor = normalizeSize ? ((maxSize == 0f) ? 1f : (float) Math.sqrt(entrySize / maxSize)) :
                entrySize;
        final float shapeSize = reference * factor;
        return shapeSize;
    }

    protected void drawDataSet(Canvas c, IBubbleDataSet dataSet) {

        if (dataSet.getEntryCount() < 1)
            return;

        Transformer trans = mChart.getTransformer(dataSet.getAxisDependency());

        float phaseY = animator.getPhaseY();

        xBounds.set(mChart, dataSet);

        sizeBuffer[0] = 0f;
        sizeBuffer[2] = 1f;

        trans.pointValuesToPixel(sizeBuffer);

        boolean normalizeSize = dataSet.isNormalizeSizeEnabled();

        // calcualte the full width of 1 step on the x-axis
        final float maxBubbleWidth = Math.abs(sizeBuffer[2] - sizeBuffer[0]);
        final float maxBubbleHeight = Math.abs(viewPortHandler.contentBottom() - viewPortHandler.contentTop());
        final float referenceSize = Math.min(maxBubbleHeight, maxBubbleWidth);

        for (int j = xBounds.min; j <= xBounds.range + xBounds.min; j++) {

            final BubbleEntry entry = dataSet.getEntryForIndex(j);

            pointBuffer[0] = entry.getX();
            pointBuffer[1] = (entry.getY()) * phaseY;
            trans.pointValuesToPixel(pointBuffer);

            float shapeHalf = getShapeSize(entry.getSize(), dataSet.getMaxSize(), referenceSize, normalizeSize) / 2f;

            if (!viewPortHandler.isInBoundsTop(pointBuffer[1] + shapeHalf)
                    || !viewPortHandler.isInBoundsBottom(pointBuffer[1] - shapeHalf))
                continue;

            if (!viewPortHandler.isInBoundsLeft(pointBuffer[0] + shapeHalf))
                continue;

            if (!viewPortHandler.isInBoundsRight(pointBuffer[0] - shapeHalf))
                break;

            final int color = dataSet.getColor(j);

            renderPaint.setColor(color);
            c.drawCircle(pointBuffer[0], pointBuffer[1], shapeHalf, renderPaint);
        }
    }

    @Override
    public void drawValues(Canvas c) {

        BubbleData bubbleData = mChart.getBubbleData();

        if (bubbleData == null)
            return;

        // if values are drawn
        if (isDrawingValuesAllowed(mChart)) {

            final List<IBubbleDataSet> dataSets = bubbleData.getDataSets();

            float lineHeight = Utils.calcTextHeight(mValuePaint, "1");

            for (int i = 0; i < dataSets.size(); i++) {

                IBubbleDataSet dataSet = dataSets.get(i);
                if (dataSet.getEntryCount() == 0) {
                    continue;
                }
                if (!shouldDrawValues(dataSet) || dataSet.getEntryCount() < 1) {
                    continue;
                }

                // apply the text-styling defined by the DataSet
                applyValueTextStyle(dataSet);

                final float phaseX = Math.max(0.f, Math.min(1.f, animator.getPhaseX()));
                final float phaseY = animator.getPhaseY();

                xBounds.set(mChart, dataSet);

                final float[] positions = mChart.getTransformer(dataSet.getAxisDependency())
                        .generateTransformedValuesBubble(dataSet, phaseY, xBounds.min, xBounds.max);

                final float alpha = phaseX == 1 ? phaseY : phaseX;

                MPPointF iconsOffset = MPPointF.getInstance(dataSet.getIconsOffset());
                iconsOffset.x = Utils.convertDpToPixel(iconsOffset.x);
                iconsOffset.y = Utils.convertDpToPixel(iconsOffset.y);

                for (int j = 0; j < positions.length; j += 2) {

                    int valueTextColor = dataSet.getValueTextColor(j / 2 + xBounds.min);
                    valueTextColor = Color.argb(Math.round(255.f * alpha), Color.red(valueTextColor),
                            Color.green(valueTextColor), Color.blue(valueTextColor));

                    float x = positions[j];
                    float y = positions[j + 1];

                    if (!viewPortHandler.isInBoundsRight(x))
                        break;

                    if ((!viewPortHandler.isInBoundsLeft(x) || !viewPortHandler.isInBoundsY(y)))
                        continue;

                    BubbleEntry entry = dataSet.getEntryForIndex(j / 2 + xBounds.min);

                    if (dataSet.isDrawValuesEnabled()) {
                        drawValue(c, dataSet.getValueFormatter(), entry.getSize(), entry, i, x,
                                y + (0.5f * lineHeight), valueTextColor);
                    }

                    if (entry.getIcon() != null && dataSet.isDrawIconsEnabled()) {

                        Drawable icon = entry.getIcon();

                        Utils.drawImage(
                                c,
                                icon,
                                (int)(x + iconsOffset.x),
                                (int)(y + iconsOffset.y),
                                icon.getIntrinsicWidth(),
                                icon.getIntrinsicHeight());
                    }
                }

                MPPointF.recycleInstance(iconsOffset);
            }
        }
    }

    @Override
    public void drawExtras(Canvas c) {
    }

    private float[] _hsvBuffer = new float[3];

    @Override
    public void drawHighlighted(Canvas c, Highlight[] indices) {

        BubbleData bubbleData = mChart.getBubbleData();

        float phaseY = animator.getPhaseY();

        for (Highlight high : indices) {

            IBubbleDataSet set = bubbleData.getDataSetByIndex(high.getDataSetIndex());

            if (set == null || !set.isHighlightEnabled())
                continue;

            final BubbleEntry entry = set.getEntryForXValue(high.getX(), high.getY());

            if (entry.getY() != high.getY())
                continue;

            if (!isInBoundsX(entry, set))
                continue;

            Transformer trans = mChart.getTransformer(set.getAxisDependency());

            sizeBuffer[0] = 0f;
            sizeBuffer[2] = 1f;

            trans.pointValuesToPixel(sizeBuffer);

            boolean normalizeSize = set.isNormalizeSizeEnabled();

            // calcualte the full width of 1 step on the x-axis
            final float maxBubbleWidth = Math.abs(sizeBuffer[2] - sizeBuffer[0]);
            final float maxBubbleHeight = Math.abs(
                    viewPortHandler.contentBottom() - viewPortHandler.contentTop());
            final float referenceSize = Math.min(maxBubbleHeight, maxBubbleWidth);

            pointBuffer[0] = entry.getX();
            pointBuffer[1] = (entry.getY()) * phaseY;
            trans.pointValuesToPixel(pointBuffer);

            high.setDraw(pointBuffer[0], pointBuffer[1]);

            float shapeHalf = getShapeSize(entry.getSize(),
                    set.getMaxSize(),
                    referenceSize,
                    normalizeSize) / 2f;

            if (!viewPortHandler.isInBoundsTop(pointBuffer[1] + shapeHalf)
                    || !viewPortHandler.isInBoundsBottom(pointBuffer[1] - shapeHalf))
                continue;

            if (!viewPortHandler.isInBoundsLeft(pointBuffer[0] + shapeHalf))
                continue;

            if (!viewPortHandler.isInBoundsRight(pointBuffer[0] - shapeHalf))
                break;

            final int originalColor = set.getColor((int) entry.getX());

            Color.RGBToHSV(Color.red(originalColor), Color.green(originalColor),
                    Color.blue(originalColor), _hsvBuffer);
            _hsvBuffer[2] *= 0.5f;
            final int color = Color.HSVToColor(Color.alpha(originalColor), _hsvBuffer);

            highlightPaint.setColor(color);
            highlightPaint.setStrokeWidth(set.getHighlightCircleWidth());
            c.drawCircle(pointBuffer[0], pointBuffer[1], shapeHalf, highlightPaint);
        }
    }
}

File: chartLib/src/main/kotlin/info/appdev/charting/renderer/BubbleChartRenderer.kt
package info.appdev.charting.renderer

import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import info.appdev.charting.animation.ChartAnimator
import info.appdev.charting.highlight.Highlight
import info.appdev.charting.interfaces.dataprovider.BubbleDataProvider
import info.appdev.charting.interfaces.datasets.IBubbleDataSet
import info.appdev.charting.utils.PointF
import info.appdev.charting.utils.Utils
import info.appdev.charting.utils.ViewPortHandler
import info.appdev.charting.utils.calcTextHeight
import info.appdev.charting.utils.convertDpToPixel
import kotlin.math.abs
import kotlin.math.max
import kotlin.math.min
import kotlin.math.roundToInt
import kotlin.math.sqrt

@Suppress("MemberVisibilityCanBePrivate")
open class BubbleChartRenderer(
    var dataProvider: BubbleDataProvider,
    animator: ChartAnimator,
    viewPortHandler: ViewPortHandler
) : BarLineScatterCandleBubbleRenderer(animator, viewPortHandler) {

    override fun initBuffers() = Unit

    override fun drawData(canvas: Canvas) {
        val bubbleData = dataProvider.bubbleData

        bubbleData?.dataSets?.forEach { set ->
            if (set.isVisible)
                drawDataSet(canvas, set)
        }
    }

    private val sizeBuffer = FloatArray(4)
    private val pointBuffer = FloatArray(2)

    protected fun getShapeSize(entrySize: Float, maxSize: Float, reference: Float, normalizeSize: Boolean): Float {
        val factor = if (normalizeSize) (if (maxSize == 0f) 1f else sqrt((entrySize / maxSize).toDouble()).toFloat()) else entrySize
        val shapeSize = reference * factor
        return shapeSize
    }

    protected fun drawDataSet(canvas: Canvas, dataSet: IBubbleDataSet) {
        if (dataSet.entryCount < 1)
            return

        val trans = dataProvider.getTransformer(dataSet.axisDependency)

        val phaseY = animator.phaseY

        xBounds.set(dataProvider, dataSet)

        sizeBuffer[0] = 0f
        sizeBuffer[2] = 1f

        trans!!.pointValuesToPixel(sizeBuffer)

        val normalizeSize = dataSet.isNormalizeSizeEnabled

        // calculate the full width of 1 step on the x-axis
        val maxBubbleWidth = abs((sizeBuffer[2] - sizeBuffer[0]).toDouble()).toFloat()
        val maxBubbleHeight = abs((viewPortHandler.contentBottom() - viewPortHandler.contentTop()).toDouble()).toFloat()
        val referenceSize = min(maxBubbleHeight.toDouble(), maxBubbleWidth.toDouble()).toFloat()

        for (j in xBounds.min..xBounds.range + xBounds.min) {
            val bubbleEntry = dataSet.getEntryForIndex(j)
            bubbleEntry?.let {
                pointBuffer[0] = it.x
                pointBuffer[1] = (it.y) * phaseY
                trans.pointValuesToPixel(pointBuffer)

                val shapeHalf = getShapeSize(it.size, dataSet.maxSize, referenceSize, normalizeSize) / 2f

                if (!viewPortHandler.isInBoundsTop(pointBuffer[1] + shapeHalf)
                    || !viewPortHandler.isInBoundsBottom(pointBuffer[1] - shapeHalf)
                ) continue

                if (!viewPortHandler.isInBoundsLeft(pointBuffer[0] + shapeHalf)) continue

                if (!viewPortHandler.isInBoundsRight(pointBuffer[0] - shapeHalf)) break

                val color = dataSet.getColorByIndex(j)

                paintRender.color = color
                canvas.drawCircle(pointBuffer[0], pointBuffer[1], shapeHalf, paintRender)
            }
        }
    }

    override fun drawValues(canvas: Canvas) {
        val bubbleData = dataProvider.bubbleData ?: return

        // if values are drawn
        if (isDrawingValuesAllowed(dataProvider)) {
            val dataSets = bubbleData.dataSets

            val lineHeight = paintValues.calcTextHeight("1").toFloat()

            dataSets?.let {
                for (i in it.indices) {
                    val dataSet = it[i]
                    if (dataSet.entryCount == 0) {
                        continue
                    }
                    if (!shouldDrawValues(dataSet) || dataSet.entryCount < 1) {
                        continue
                    }

                    // apply the text-styling defined by the DataSet
                    applyValueTextStyle(dataSet)

                    val phaseX = max(0.0, min(1.0, animator.phaseX.toDouble())).toFloat()
                    val phaseY = animator.phaseY

                    xBounds.set(dataProvider, dataSet)

                    dataProvider.getTransformer(dataSet.axisDependency)?.let { transformer ->
                        val positions = transformer.generateTransformedValuesBubble(dataSet, phaseY, xBounds.min, xBounds.max)

                        val alpha = if (phaseX == 1f)
                            phaseY
                        else
                            phaseX

                        val iconsOffset = PointF.getInstance(dataSet.iconsOffset)
                        iconsOffset.x = iconsOffset.x.convertDpToPixel()
                        iconsOffset.y = iconsOffset.y.convertDpToPixel()

                        var j = 0
                        while (j < positions.size) {
                            var valueTextColor = dataSet.getValueTextColor(j / 2 + xBounds.min)
                            valueTextColor = Color.argb(
                                (255f * alpha).roundToInt(), Color.red(valueTextColor),
                                Color.green(valueTextColor), Color.blue(valueTextColor)
                            )

                            val x = positions[j]
                            val y = positions[j + 1]

                            if (!viewPortHandler.isInBoundsRight(x)) break

                            if ((!viewPortHandler.isInBoundsLeft(x) || !viewPortHandler.isInBoundsY(y))) {
                                j += 2
                                continue
                            }

                            val bubbleEntry = dataSet.getEntryForIndex(j / 2 + xBounds.min)
                            bubbleEntry?.let {
                                if (dataSet.isDrawValues) {
                                    drawValue(
                                        canvas, dataSet.valueFormatter, bubbleEntry.size, bubbleEntry, i, x,
                                        y + (0.5f * lineHeight), valueTextColor
                                    )
                                }

                                if (bubbleEntry.icon != null && dataSet.isDrawIcons) {
                                    val icon = bubbleEntry.icon

                                    icon?.let { ico ->
                                        Utils.drawImage(
                                            canvas,
                                            ico,
                                            (x + iconsOffset.x).toInt(),
                                            (y + iconsOffset.y).toInt()
                                        )
                                    }
                                }
                            }
                            j += 2
                        }

                        PointF.recycleInstance(iconsOffset)
                    }
                }
            }
        }
    }

    override fun drawExtras(canvas: Canvas) {
    }

    private val _hsvBuffer = FloatArray(3)

    init {
        paintRender.style = Paint.Style.FILL

        paintHighlight.style = Paint.Style.STROKE
        paintHighlight.strokeWidth = 1.5f.convertDpToPixel()
    }

    override fun drawHighlighted(canvas: Canvas, indices: Array<Highlight>) {
        val bubbleData = dataProvider.bubbleData

        val phaseY = animator.phaseY

        for (high in indices) {
            val set = bubbleData?.getDataSetByIndex(high.dataSetIndex)

            if (set == null || !set.isHighlightEnabled)
                continue

            val bubbleEntry = set.getEntryForXValue(high.x, high.y)!!

            if (bubbleEntry.y != high.y)
                continue

            if (!isInBoundsX(bubbleEntry, set))
                continue

            val trans = dataProvider.getTransformer(set.axisDependency)

            sizeBuffer[0] = 0f
            sizeBuffer[2] = 1f

            trans!!.pointValuesToPixel(sizeBuffer)

            val normalizeSize = set.isNormalizeSizeEnabled

            // calculate the full width of 1 step on the x-axis
            val maxBubbleWidth = abs((sizeBuffer[2] - sizeBuffer[0]).toDouble()).toFloat()
            val maxBubbleHeight = abs((viewPortHandler.contentBottom() - viewPortHandler.contentTop()).toDouble()).toFloat()
            val referenceSize = min(maxBubbleHeight.toDouble(), maxBubbleWidth.toDouble()).toFloat()

            pointBuffer[0] = bubbleEntry.x
            pointBuffer[1] = (bubbleEntry.y) * phaseY
            trans.pointValuesToPixel(pointBuffer)

            high.setDraw(pointBuffer[0], pointBuffer[1])

            val shapeHalf = getShapeSize(
                bubbleEntry.size,
                set.maxSize,
                referenceSize,
                normalizeSize
            ) / 2f

            if (!viewPortHandler.isInBoundsTop(pointBuffer[1] + shapeHalf)
                || !viewPortHandler.isInBoundsBottom(pointBuffer[1] - shapeHalf)
            ) continue

            if (!viewPortHandler.isInBoundsLeft(pointBuffer[0] + shapeHalf))
                continue

            if (!viewPortHandler.isInBoundsRight(pointBuffer[0] - shapeHalf))
                break

            val originalColor = set.getColorByIndex(bubbleEntry.x.toInt())

            Color.RGBToHSV(
                Color.red(originalColor), Color.green(originalColor),
                Color.blue(originalColor), _hsvBuffer
            )
            _hsvBuffer[2] *= 0.5f
            val color = Color.HSVToColor(Color.alpha(originalColor), _hsvBuffer)

            paintHighlight.color = color
            paintHighlight.strokeWidth = set.highlightCircleWidth
            canvas.drawCircle(pointBuffer[0], pointBuffer[1], shapeHalf, paintHighlight)
        }
    }
}
